feat(stats): add C implementation for `stats/base/dists/chisquare/mgf`

anandkaranubc opened the pull request:
### Resolves  
https://github.com/stdlib-js/stdlib/issues/3506  

---

### Description  

#### Purpose of this pull request  
This pull request:  
- Adds a C implementation of the moment-generating function (MGF) for a chi-squared distribution.  
- Includes benchmarks for the C implementation.  
- Provides examples demonstrating the usage of the moment-generating function (MGF) for a chi-squared distribution in C.  

---

### Related Issues  

#### Does this pull request have any related issues?  
This pull request:  
- Resolves [Issue #3506](https://github.com/stdlib-js/stdlib/issues/3506).  

---

### Questions  

#### Any questions for reviewers of this pull request?  
No.  

---

### Other Information  

#### Any other relevant information?  
No.  

---

### Checklist  

Please ensure the following tasks are completed before submitting this pull request:  
- [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md).  

---  
**@stdlib-js/reviewers**

anandkaranubc committed:
feat: add the C src files

anandkaranubc committed:
feat: add the C benchmark and example files

anandkaranubc committed:
feat: add the JS native files

anandkaranubc committed:
build: added the json and gyp dependency files

anandkaranubc committed:
docs: update README with C API documentation

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
/stdlib lint-autofix

anandkaranubc commented:
@Planeshifter @kgryte 

Thank you for the OH. I was wondering if I can get some feedback from you on this PR. This is my most recent PR of all the ones I have created, based on learning from previous ones. I hope I can use the feedback I get on this PR and make any necessary changes in my other PRs as well based on this one. Thank you so much!

Planeshifter commented:
@anandkaranubc Taking a look now...

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/README.md:
```diff
@@ -151,6 +151,106 @@ for ( i = 0; i < 10; i++ ) {
+
+        result = stdlib_base_dists_chisquare_mgf( t, k );
+
+        printf( "t: %lf, k: %lf, MGF: %lf \n", t, k, result );
```
See JS example:

```suggestion
        printf( "t: %lf, k: %lf, M_X(t;k): %lf \n", t, k, result );
```

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+
+        result = stdlib_base_dists_chisquare_mgf( t, k );
+
+        printf( "t: %lf, k: %lf, MGF: %lf \n", t, k, result );
```
```suggestion
        printf( "t: %lf, k: %lf, M_X(t;k): %lf \n", t, k, result );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+}
+
+int main( void ) {
+    double t;
+    double k;
+    double result;
+    int i;
```
By convention, we order variable declarations by length:

```suggestion
    double result;
    double t;
    double k;
    int i;
```

NishchayRajput replied:
thanks for mentioning! Was checking the other PRs comments so that I may fix the similar mistakes in mine! ehe!

Planeshifter commented:
/stdlib lint-autofix

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/package.json:
```diff
@@ -14,11 +14,14 @@
   "directories": {
     "benchmark": "./benchmark",
     "doc": "./docs",
+    "include": "./include",
     "example": "./examples",
```
Ordered alphabetically to make the CI linter happy: 

```suggestion
    "example": "./examples",
    "include": "./include",
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+#include <stdio.h>
+
+static double random_uniform( const double min, const double max ) {
+    double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
```
This file should be tab-indented. We had a bug  on CI that caused the EditorConfig linting to not catch these. Make sure you have EditorConfig setup in your IDE; then saving C files should automatically indent them using TABs

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/binding.gyp:
```diff
@@ -0,0 +1,149 @@
+# [1]: https://gyp.gsrc.io/docs/InputFormatReference.md
+# [2]: https://gyp.gsrc.io/docs/UserDocumentation.md
+{
+    # List of files to include in this file:
```
Should be two-space indented. Should be correctly formatted on save when EditorConfig is setup.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/include.gypi:
```diff
@@ -0,0 +1,48 @@
+# [1]: https://gyp.gsrc.io/docs/InputFormatReference.md
+# [2]: https://gyp.gsrc.io/docs/UserDocumentation.md
+{
+    # Define variables to be used throughout the configuration for all targets:
```
Two-space indentation.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/src/main.c:
```diff
@@ -0,0 +1,47 @@
+* @param k    degrees of freedom (must be non-negative)
+* @return     evaluated MGF, or NaN if input is invalid
+*
+* @example
+* double y = stdlib_base_dists_chisquare_mgf( 0.4, 2.0 );
+* // returns ~5.0
+*
+* @example
+* double y = stdlib_base_dists_chisquare_mgf( -1.0, 5.0 );
+* // returns ~0.0642
+*
+* @example
+* double y = stdlib_base_dists_chisquare_mgf( 0.0, 10.0 );
+* // returns 1.0
+*/
```
Per convention, we only keep one example in the C files:

```suggestion
* @example
* double y = stdlib_base_dists_chisquare_mgf( 0.4, 2.0 );
* // returns ~5.0
*/
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/src/main.c:
```diff
@@ -0,0 +1,47 @@
+* // returns 1.0
+*/
+double stdlib_base_dists_chisquare_mgf( const double t, const double k ) {
+    if ( stdlib_base_is_nan( t ) || stdlib_base_is_nan( k ) || k < 0.0 || t >= 0.5 ) {
```
Should be TAB-indentation here. You could also break up the if-clause over multiple lines.

anandkaranubc replied:
All the changes have been merged successfully! 😃

Planeshifter reviewed (commented):
Overall the PR looks good, but there are a bunch of formatting errors that need to be fixed. Make sure you have `EditorConfig` setup correctly.

Looking forward to merge your first PR soon (of many already... 😅 )! 🚀

anandkaranubc commented:
> Overall the PR looks good, but there are a bunch of formatting errors that need to be fixed. Make sure you have `EditorConfig` setup correctly.
> 
> Looking forward to merge your first PR soon (of many already... 😅 )! 🚀

Thank you so much @Planeshifter for your time and detailed feedback. I feel very bad to have overlooked all these details. Turning all my PRs to draft and making the changes as you mentioned.

anandkaranubc committed:
docs: update README with proper print info in examples

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
refactor: formatting c files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
refactor: formatting dependency files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
/stdlib lint-autofix

anandkaranubc commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

anandkaranubc committed:
refactor: formatting benchmark.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
/stdlib lint-autofix

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/README.md:
```diff
@@ -151,6 +151,105 @@ for ( i = 0; i < 10; i++ ) {
+}
+
+int main( void ) {
+    double t;
+    double k;
+    double result;
+    int i;
```
```suggestion
    double result;
    double t;
    double k;
    int i;
```

Planeshifter committed:
docs: reorder declarations in example code

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/src/main.c:
```diff
@@ -0,0 +1,39 @@
+#include "stdlib/math/base/special/pow.h"
+
+/**
+* Evaluates the moment-generating function (MGF) for a chi-squared distribution
+* with degrees of freedom `k` at a value `t`.
```
```suggestion
* Evaluates the moment-generating function (MGF) for a chi-squared distribution with degrees of freedom `k` at a value `t`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+	double k;
+	int i;
+
+	for ( i = 0; i < 10; i++ ) {
+		t = random_uniform( -0.5, 0.4 );
+		k = random_uniform( 0.1, 10.0 );
+
+		result = stdlib_base_dists_chisquare_mgf( t, k );
+
+		printf( "t: %lf, k: %lf, M_X(t;k): %lf \n", t, k, result );
+	}
+
+	return 0;
```
```suggestion
	for ( i = 0; i < 10; i++ ) {
		t = random_uniform( -0.5, 0.4 );
		k = random_uniform( 0.1, 10.0 );
		result = stdlib_base_dists_chisquare_mgf( t, k );
		printf( "t: %lf, k: %lf, M_X(t;k): %lf \n", t, k, result );
	}
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/chisquare/mgf/README.md:
```diff
@@ -151,6 +151,105 @@ for ( i = 0; i < 10; i++ ) {
+    double k;
+    int i;
+
+    for ( i = 0; i < 10; i++ ) {
+        t = random_uniform( -0.5, 0.4 );
+        k = random_uniform( 0.1, 10.0 );
+
+        result = stdlib_base_dists_chisquare_mgf( t, k );
+
+        printf( "t: %lf, k: %lf, M_X(t;k): %lf \n", t, k, result );
+    }
+
+    return 0;
```
```suggestion
    for ( i = 0; i < 10; i++ ) {
        t = random_uniform( -0.5, 0.4 );
        k = random_uniform( 0.1, 10.0 );
        result = stdlib_base_dists_chisquare_mgf( t, k );
        printf( "t: %lf, k: %lf, M_X(t;k): %lf \n", t, k, result );
    }
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Thanks for your updates! PR should be in good shape now. Will merge shortly after CI checks have passed.
